install.sh

```diff
@@ -0,0 +1,472 @@
+  systemctl --user enable --now code-server
+Or, if you don't want/need a background service you can run:
+  code-server
+EOF
+}
+
+main() {
+  if [ "${TRACE-}" ]; then
+    set -x
+  fi
+
+  unset \
+    DRY_RUN \
+    METHOD \
+    ARCHIVE_INSTALL_PREFIX \
+    SKIP_ECHO \
+    VERSION \
+    OPTIONAL
+
+  while [ "$#" -gt 0 ]; do
+    case "$1" in
+    --dry-run)
+      DRY_RUN=1
+      ;;
+    --method)
+      METHOD="$(parse_arg "$@")"
+      shift
+      ;;
+    --method=*)
+      METHOD="$(parse_arg "$@")"
+      ;;
+    --prefix)
+      ARCHIVE_INSTALL_PREFIX="$(parse_arg "$@")"
+      shift
+      ;;
+    --prefix=*)
+      ARCHIVE_INSTALL_PREFIX="$(parse_arg "$@")"
+      ;;
+    --version)
+      VERSION="$(parse_arg "$@")"
+      shift
+      ;;
+    --version=*)
+      VERSION="$(parse_arg "$@")"
+      ;;
+    -h | --h | -help | --help)
+      usage
+      exit 0
+      ;;
+    *)
+      echoerr "Unknown flag $1"
+      echoerr "Run with --help to see usage."
+      exit 1
+      ;;
+    esac
+
+    shift
+  done
+
+  VERSION="${VERSION-$(echo_latest_version)}"
+  METHOD="${METHOD-detect}"
+  if [ "$METHOD" != detect ] && [ "$METHOD" != archive ]; then
+    echoerr "Unknown install method \"$METHOD\""
+    echoerr "Run with --help to see usage."
+    exit 1
+  fi
+  ARCHIVE_INSTALL_PREFIX="${ARCHIVE_INSTALL_PREFIX-$HOME/.local}"
+
+  OS="$(os)"
+  if [ ! "$OS" ]; then
+    echoerr "Unsupported OS $(uname)."
+    exit 1
+  fi
+
+  distro_name
+
+  ARCH="$(arch)"
+  if [ ! "$ARCH" ]; then
+    if [ "$METHOD" = archive ]; then
+      echoerr "No static releases available for the architecture $(uname -m)."
+      echoerr 'Please rerun without the "--method archive" flag to install from npm.'
+      exit 1
+    fi
+    echo "No precompiled releases for $(uname -m)."
+    install_npm
+    return
+  fi
+
+  CACHE_DIR="$(echo_cache_dir)"
+  mkdir -p "$CACHE_DIR"
+
+  if [ "$METHOD" = archive ]; then
+    install_archive
+    return
+  fi
+
+  case "$(distro)" in
+  macos)
+    install_macos
+    ;;
+  ubuntu | debian | raspbian)
+    install_deb
+    ;;
+  centos | fedora | rhel | opensuse)
+    install_rpm
+    ;;
+  arch)
+    install_aur
+    ;;
+  *)
+    echo "Unsupported package manager."
+    install_archive
+    ;;
+  esac
+}
+
+parse_arg() {
+  case "$1" in
+  *=*)
+    opt="${1#=*}"
+    optarg="${1#*=}"
+    if [ ! "$optarg" ] && [ ! "${OPTIONAL-}" ]; then
+      echoerr "$opt requires an argument"
+      echoerr "Run with --help to see usage."
+      exit 1
+    fi
+    echo "$optarg"
+    return
+    ;;
+  esac
+
+  case "${2-}" in
+  "" | -*)
+    if [ ! "${OPTIONAL-}" ]; then
+      echoerr "$1 requires an argument"
+      echoerr "Run with --help to see usage."
+      exit 1
+    fi
+    ;;
+  *)
+    echo "$2"
+    return
+    ;;
+  esac
+}
+
+fetch() {
+  URL="$1"
+  FILE="$2"
+
+  echo "+ Downloading $URL"
+
+  if [ -e "$FILE" ]; then
+    echo "+ Using cached $FILE"
+    return
+  fi
+
+  SKIP_ECHO=1
```

Comment:
Should we leave out the `SKIP_ECHO` here and elsewhere? It's a little unfortunate that when I use `--dry-run` I don't see the actual commands being ran. 

Or we could update the description of `--dry-run` to say it lists steps and not necessarily commands. 

Or if `--dry-run` is passed then ignore the skip echo flag.

Reply:
So this was intentional as the commands can be quite verbose. That's why I added some logs to explain what's being done. E.g. the `Downloading` log.

Reply:
That's fair, I mostly just feel like the current description of `--dry-run` is a bit misleading.

fwiw I thought being able to see the actual commands with `--dry-run` was pretty dope since you can fully audit what's going to happen. Skipping some commands somewhat lessons the usefulness of the flag imo (at least, in the way I was expecting to use it).

Reply:
`--dry-run` prints out the process, not the commands. I don't think printing out the exact commands is useful when they're really verbose, like the curl's take up multiple lines.

If you really want the commands for tracing, you can use `TRACE=1`.

Reply:
That's true but I find `TRACE=1` really difficult to follow.

My impression was that `--dry-run` was a way for you to audit the script before running it but you can't fully audit the script if some commands are obfuscated, right?

I think part of the reason I have that impression is that the description says `Echo the commands` so if we change that to something like `Echo the steps` that might do it.

Reply:
My bad, I thought it already said echo the process, will fix and make the logs a little more clear.

Reply:
Ended up removing `SKIP_ECHO` and printing all install commands and piping all output through a sed that replaces `$HOME` with `~`. Removes the verbosity enough.